File: android/apps/ensu/app/src/main/java/io/ente/ensu/notes/NotesStore.kt
package io.ente.ensu.notes

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.compose.runtime.staticCompositionLocalOf
import io.ente.ensu.bindings.GroundedExcerpt
import io.ente.ensu.bindings.GroundedSource
import io.ente.ensu.bindings.NoteSourceReference
import io.ente.ensu.bindings.NotesCancellation
import io.ente.ensu.bindings.NotesException
import io.ente.ensu.bindings.NotesHit
import io.ente.ensu.bindings.NotesIndexOptions
import io.ente.ensu.bindings.NotesProgress
import io.ente.ensu.bindings.NotesProgressCallback
import io.ente.ensu.bindings.NotesSummary
import io.ente.ensu.bindings.notesLimits
import io.ente.ensu.bindings.withNotesCollectionLabel
import io.ente.ensu.llm.LlmProvider
import io.ente.ensu.llm.ModelMaintenance
import io.ente.ensu.llm.withMaintenanceSuspended
import io.ente.ensu.logging.FileLogRepository
import io.ente.ensu.logging.LogLevel
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import java.util.concurrent.atomic.AtomicBoolean

enum class NotesStatus { Pending, Indexing, Updating, Ready, Unavailable, Error }

data class NoteCollectionState(
    val id: String,
    val label: String,
    val status: NotesStatus = NotesStatus.Pending,
    val documentCount: Long = 0,
    val indexAvailable: Boolean = false,
    val progress: Int? = null,
    val error: String? = null,
    val completedEmpty: Boolean = false,
    val lastUpdatedAtMs: Long? = null
) {
    val eligible: Boolean
        get() = indexAvailable && status != NotesStatus.Error && status != NotesStatus.Unavailable
}

data class NotesState(
    val collections: List<NoteCollectionState> = emptyList(),
    val error: String? = null
)

val LocalNotesStore = staticCompositionLocalOf<NotesStore?> { null }

class NotesStore(
    private val context: Context,
    private val llm: LlmProvider,
    private val logRepository: FileLogRepository,
    private val provider: NotesProvider = NotesProvider(context)
) : ModelMaintenance {
    private data class Pending(
        val forced: Set<String>,
        val rebuild: Boolean,
        val due: Long
    )

    private class Run(
        val cancel: NotesCancellation,
        val job: Job
    )

    private val limits = notesLimits()
    private val registrations = linkedMapOf<String, NotesRegistration>()
    private val pending = linkedMapOf<String, Pending>()
    private val scans = linkedSetOf<String>()
    private val failed = mutableSetOf<String>()
    private val maintenance = NotesMaintenanceGate<Run>()
    private var scope: CoroutineScope? = null
    private var bootstrapJob: Job? = null
    private val active get() = maintenance.active
    private var wake: Job? = null
    private var foreground = false
    private var loaded = false
    private var disabled = false
    private val _state = MutableStateFlow(NotesState())
    val state: StateFlow<NotesState> = _state.asStateFlow()

    fun bootstrap(owner: CoroutineScope) {
        if (bootstrapJob != null) return
        scope = owner
        bootstrapJob = owner.launch {
            try {
                val records = provider.load()
                registrations.putAll(records.associateBy { it.id })
                val states = records.map { record ->
                    try {
                        val summary = provider.inspect(record.id)
                        if (!summary.initialComplete) enqueue(record.id, immediate = true)
                        fromSummary(record, summary)
                    } catch (e: NotesException.RebuildRequired) {
                        enqueue(record.id, rebuild = true, immediate = true)
                        NoteCollectionState(record.id, record.label)
                    } catch (e: Exception) {
                        logRepository.log(LogLevel.Error, "Checkpoint load failed collection=${record.id}", tag = "Notes", throwable = e)
                        NoteCollectionState(record.id, record.label, NotesStatus.Error, error = message(e))
                    }
                }
                _state.value = NotesState(states)
                loaded = true
                scans.addAll(registrations.keys)
            } catch (e: Exception) {
                logRepository.log(LogLevel.Error, "Collection registry load failed", tag = "Notes", throwable = e)
                disabled = true
                _state.value = NotesState(error = "Could not load Your Notes. Existing indexes have been preserved.")
            }
            pump()
        }
    }

    suspend fun awaitReady() {
        bootstrapJob?.join()
    }

    fun setForeground(value: Boolean) {
        if (foreground == value) return
        foreground = value
        if (!value) {
            wake?.cancel()
            active?.job?.cancel()
        } else {
            failed.clear()
            if (loaded) scans.addAll(registrations.keys)
            pump()
        }
    }

    override fun modelReadinessChanged() {
        pump()
    }

    override fun suspendMaintenance(): AutoCloseable {
        maintenance.suspend()?.job?.cancel()
        val closed = AtomicBoolean()
        return AutoCloseable {
            if (closed.compareAndSet(false, true)) {
                maintenance.resume()
                scope?.launch { pump() }
            }
        }
    }

    override suspend fun awaitMaintenance() {
        active?.job?.join()
    }

    fun add(tree: Uri) {
        scope?.launch {
            awaitReady()
            if (disabled) return@launch
            withMaintenanceSuspended {
                try {
                    val record = provider.add(tree)
                    registrations[record.id] = record
                    _state.update {
                        it.copy(collections = it.collections + NoteCollectionState(record.id, record.label), error = null)
                    }
                    enqueue(record.id, immediate = true)
                } catch (error: CancellationException) {
                    throw error
                } catch (error: Exception) {
                    _state.update { it.copy(error = message(error)) }
                }
            }
        }
    }

    fun remove(id: String) {
        scope?.launch {
            withMaintenanceSuspended {
                if (id !in registrations) return@withMaintenanceSuspended
                try {
                    provider.remove(id)
                    registrations.remove(id)
                    pending.remove(id)
                    scans.remove(id)
                    failed.remove(id)
                    _state.update {
                        it.copy(collections = it.collections.filter { row -> row.id != id }, error = null)
                    }
                } catch (error: CancellationException) {
                    throw error
                } catch (error: Exception) {
                    _state.update { it.copy(error = message(error)) }
                }
            }
        }
    }

    fun retry(id: String) {
        if (!loaded || disabled || id !in registrations) return
        failed.remove(id)
        pending[id]?.let { pending[id] = it.copy(due = 0) }
        scans.add(id)
        update(id) { it.copy(status = NotesStatus.Pending, progress = null, error = null) }
        pump()
    }

    private fun clearError() {
        _state.update { it.copy(error = null) }
    }

    suspend fun retrieve(query: List<Float>): List<NotesHit> = withContext(Dispatchers.Main.immediate) {
        val records = _state.value.collections.filter { it.eligible }.mapNotNull { registrations[it.id] }
        val hits = mutableListOf<NotesHit>()
        for (record in records) {
            currentCoroutineContext().ensureActive()
            try {
                hits += provider.search(record.id, query)
            } catch (error: CancellationException) {
                throw error
            } catch (error: NotesException.RebuildRequired) {
                enqueue(record.id, rebuild = true, immediate = true)
                update(record.id) { it.copy(status = NotesStatus.Pending, indexAvailable = false) }
            } catch (error: Exception) {
                fail(record.id, error)
            }
        }
        hits
    }

    suspend fun verify(excerpts: List<GroundedExcerpt>): List<GroundedExcerpt> =
        withContext(Dispatchers.Main.immediate) {
            val records = registrations.toMap()
            var accepted = 0
            val result = mutableListOf<GroundedExcerpt>()
            for (excerpt in excerpts) {
                currentCoroutineContext().ensureActive()
                val note = (excerpt.source as? GroundedSource.LocalNote)?.reference
                if (note == null) {
                    result += excerpt
                    continue
                }
                if (accepted.toUInt() >= limits.maxGroundingHits) continue
                val record = records[note.collectionId] ?: continue
                try {
                    provider.verify(note)
                    result += excerpt.copy(source = GroundedSource.LocalNote(withNotesCollectionLabel(note, record.label)))
                    accepted++
                } catch (error: CancellationException) {
                    throw error
                } catch (error: NotesException.SourceChanged) {
                    enqueue(record.id, forced = setOf(note.documentId), immediate = true)
                } catch (error: Exception) {
                    fail(record.id, error)
                }
            }
            result
        }

    fun open(reference: NoteSourceReference) {
        scope?.launch {
            clearError()
            try {
                val uri = withMaintenanceSuspended {
                    provider.preview(reference)
                }
                context.startActivity(Intent(Intent.ACTION_VIEW).apply {
                    setDataAndType(uri, "text/plain")
                    addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_ACTIVITY_NEW_TASK)
                })
                clearError()
            } catch (e: Exception) {
                val error = when (e) {
                    is NotesException.SourceChanged -> "This note changed after the answer was generated."
                    is NotesException.Unavailable -> "This note is unavailable."
                    else -> "Could not open this note. Check folder access and that a text viewer is installed."
                }
                _state.update { it.copy(error = error) }
            }
        }
    }

    private fun enqueue(
        id: String,
        rebuild: Boolean = false,
        forced: Set<String> = emptySet(),
        immediate: Boolean = false
    ) {
        if (id !in registrations) return
        val old = pending[id]
        pending[id] = Pending(
            forced = old?.forced.orEmpty() + forced,
            rebuild = rebuild || old?.rebuild == true,
            due = if (immediate) 0 else old?.due ?: (System.currentTimeMillis() + SCAN_INTERVAL_MS)
        )
    }

    private fun pump() {
        val owner = scope ?: return
        if (!owner.isActive || !loaded || disabled || !foreground || !maintenance.available) return
        wake?.cancel()
        val now = System.currentTimeMillis()
        val embeddingReady = llm.isEmbeddingModelReady()
        val request = if (embeddingReady) {
            pending.entries.firstOrNull { it.key !in failed && it.value.due <= now }
        } else null
        val id = request?.key ?: scans.firstOrNull { it !in pending && it !in failed }
        if (id == null) {
            val due = if (embeddingReady) {
                pending.asSequence().filter { it.key !in failed }.minOfOrNull { it.value.due }
            } else null
            wake = owner.launch {
                delay(minOf(SCAN_INTERVAL_MS, due?.minus(now)?.coerceAtLeast(1) ?: SCAN_INTERVAL_MS))
                scans.addAll(_state.value.collections.filter {
                    it.status != NotesStatus.Unavailable && (it.status != NotesStatus.Error || it.completedEmpty)
                }.map { it.id })
                pump()
            }
            return
        }
        val record = registrations[id] ?: run {
            pending.remove(id)
            scans.remove(id)
            pump()
            return
        }
        val snapshot = request?.value
        val cancel = NotesCancellation()
        val hadIndex = _state.value.collections.firstOrNull { it.id == id }?.indexAvailable == true
        val job = owner.launch(start = CoroutineStart.LAZY) {
            scans.remove(id)
            try {
                cancel.check()
                if (snapshot == null) {
                    val result = provider.inspectFreshness(id, cancel)
                    update(id) { fromSummary(record, result.summary).copy(progress = if (result.summary.initialComplete) null else it.progress) }
                    if (result.changed) {
                        enqueue(id, forced = result.forcedDocumentIds.toSet(), immediate = !result.summary.initialComplete)
                        update(id) { it.copy(status = NotesStatus.Pending, error = null) }
                    }
                } else {
                    update(id) { it.copy(status = if (hadIndex) NotesStatus.Updating else NotesStatus.Indexing, error = null) }
                    val result = llm.withEmbeddingContext(checkCancellation = cancel::check) { embedding ->
                        provider.index(
                            id,
                            embedding,
                            cancel,
                            object : NotesProgressCallback {
                                override fun onProgress(progress: NotesProgress) {
                                    owner.launch {
                                        if (active?.cancel === cancel) {
                                            update(id) {
                                                it.copy(progress = progress.percentage.toInt(), documentCount = progress.indexedDocumentCount.toLong())
                                            }
                                        }
                                    }
                                }
                            },
                            NotesIndexOptions(snapshot.forced.toList(), snapshot.rebuild)
                        )
                    }
                    val changedId = result.changedDocumentId
                    if (pending[id] === snapshot) {
                        if (changedId == null) {
                            pending.remove(id)
                        } else {
                            pending[id] = snapshot.copy(
                                forced = result.uncheckedDocumentIds.toSet() + changedId,
                                rebuild = false,
                                due = System.currentTimeMillis() + SCAN_INTERVAL_MS
                            )
                        }
                    } else if (changedId != null) {
                        enqueue(id, forced = result.uncheckedDocumentIds.toSet() + changedId)
                    }
                    update(id) {
                        if (id !in pending) {
                            fromSummary(record, result.summary)
                        } else {
                            it.copy(
                                status = NotesStatus.Pending,
                                documentCount = result.summary.documentCount.toLong(),
                                indexAvailable = hadIndex && result.summary.initialComplete && result.summary.documentCount > 0uL
                            )
                        }
                    }
                }
            } catch (e: Exception) {
                when (e) {
                    is NotesException.Cancelled, is CancellationException -> {
                        if (snapshot == null) scans.add(id)
                        val saved = withContext(NonCancellable) {
                            runCatching { provider.inspect(id) }.getOrNull()
                        }
                        update(id) {
                            it.copy(
                                status = NotesStatus.Pending,
                                indexAvailable = hadIndex && saved?.initialComplete == true && saved.documentCount > 0uL
                            )
                        }
                        if (e is CancellationException) throw e
                    }
                    is NotesException.RebuildRequired -> {
                        if (snapshot == null) {
                            enqueue(id, rebuild = true, immediate = true)
                            update(id) { it.copy(status = NotesStatus.Pending, indexAvailable = false) }
                        } else fail(id, e)
                    }
                    else -> fail(id, e)
                }
            }
        }
        val run = Run(cancel, job)
        val admitted = maintenance.admit(run)
        job.invokeOnCompletion {
            cancel.destroy()
            maintenance.finish(run)
            if (admitted) owner.launch { pump() }
        }
        if (admitted) job.start() else job.cancel()
    }

    private fun fail(id: String, e: Exception) {
        if (id !in registrations) return
        logRepository.log(LogLevel.Error, "Notes update failed collection=$id", tag = "Notes", throwable = e)
        failed.add(id)
        scans.remove(id)
        update(id) {
            it.copy(
                status = if (e is NotesException.Unavailable) NotesStatus.Unavailable else NotesStatus.Error,
                progress = null,
                error = message(e),
                completedEmpty = false
            )
        }
    }

    private fun update(id: String, transform: (NoteCollectionState) -> NoteCollectionState) {
        _state.update { it.copy(collections = it.collections.map { row -> if (row.id == id) transform(row) else row }) }
    }

    private fun fromSummary(record: NotesRegistration, summary: NotesSummary): NoteCollectionState {
        val ready = summary.initialComplete && summary.documentCount > 0uL
        val empty = summary.initialComplete && !ready
        return NoteCollectionState(
            id = record.id,
            label = record.label,
            status = when {
                ready -> NotesStatus.Ready
                empty -> NotesStatus.Error
                else -> NotesStatus.Pending
            },
            documentCount = summary.documentCount.toLong(),
            indexAvailable = ready,
            error = if (empty) EMPTY_NOTES else null,
            completedEmpty = empty,
            lastUpdatedAtMs = summary.lastUpdatedAtMs
        )
    }

    private fun message(e: Exception): String = when (e) {
        is NotesException.Unavailable -> "This Notes folder is unavailable. Check access and try again."
        is NotesException.SourceChanged -> "The folder changed. Try again after it finishes updating."
        is NotesException.SourceRead -> e.detail
        is NotesException.InvalidInput -> e.detail
        is NotesException.Storage -> "Could not save the Notes index. Check storage and try again."
        is NotesException.RebuildRequired -> "The Notes index needs rebuilding."
        else -> "Could not update Your Notes. Please try again."
    }

    companion object {
        private const val SCAN_INTERVAL_MS = 300_000L
        private const val EMPTY_NOTES = "No supported non-empty UTF-8 notes were found"
    }
}

File: apple/apps/ensu/Ensu/Services/NotesStore.swift
import Foundation
import SwiftUI

enum NotesStatus {
    case pending, indexing, updating, ready, unavailable, error
}

struct NoteCollectionState: Identifiable {
    let id: String
    let label: String
    var status: NotesStatus = .pending
    var documentCount: UInt64 = 0
    var lastUpdatedAtMs: Int64?
    var indexAvailable = false
    var progress: UInt8?
    var error: String?
    var completedEmpty = false
    var eligible: Bool {
        indexAvailable && status != .error && status != .unavailable
    }
}

@MainActor
final class NotesStore: ObservableObject, ModelMaintenance {
    @Published private(set) var collections: [NoteCollectionState] = []
    @Published var operationError: String?
    @Published var preview: NotesPreview?

    private struct Pending {
        let generation: UInt64
        let forced: Set<String>
        let rebuild: Bool
        let due: Date
    }
    private struct Run {
        let operation: NotesReadOperation
        let task: Task<Void, Never>
    }

    private let modelProvider: LlmProvider
    private let provider: NotesProvider
    private let limits = notesLimits()
    private var pending: [String: Pending] = [:]
    private var scans = Set<String>()
    private var failed = Set<String>()
    private var generation: UInt64 = 0
    private var active: Run?
    private var scopes = 0
    private var foreground = false
    private var loaded = false
    private var disabled = false
    private var bootstrapTask: Task<Void, Never>?
    private var wake: Task<Void, Never>?
    private static let emptyNotes = "No supported non-empty UTF-8 notes were found"

    init(provider: LlmProvider, notesProvider: NotesProvider = NotesProvider()) {
        modelProvider = provider
        self.provider = notesProvider
    }

    func bootstrap() async {
        if let bootstrapTask {
            await bootstrapTask.value
            return
        }
        let task = Task { @MainActor in
            do {
                let records = try await provider.load()
                for record in records {
                    do {
                        let summary = try await provider.inspect(record.id)
                        collections.append(Self.state(record, summary))
                        if !summary.initialComplete {
                            enqueue(record.id, due: .distantPast)
                        }
                    } catch NotesError.RebuildRequired {
                        collections.append(NoteCollectionState(id: record.id, label: record.label))
                        enqueue(record.id, rebuild: true, due: .distantPast)
                    } catch {
                        collections.append(NoteCollectionState(id: record.id, label: record.label, status: .error,
                            error: Self.message(error)))
                    }
                }
                loaded = true
                scans.formUnion(records.map(\.id))
                pump()
            } catch {
                disabled = true
                operationError = "Could not load Your Notes. Existing indexes have been preserved."
            }
        }
        bootstrapTask = task
        await task.value
    }

    func setForeground(_ value: Bool) {
        guard foreground != value else { return }
        foreground = value
        if value {
            failed.removeAll()
            scans.formUnion(collections.map(\.id))
            pump()
        } else {
            wake?.cancel()
            active?.operation.cancel()
        }
    }
    func modelReadinessChanged() {
        pump()
    }

    func suspendMaintenance() -> ModelUseScope {
        scopes += 1
        active?.operation.cancel()
        return ModelUseScope { [weak self] in
            Task { @MainActor in
                guard let self else { return }
                self.scopes -= 1
                self.pump()
            }
        }
    }
    func awaitMaintenance() async {
        await active?.task.value
    }

    func add(_ url: URL) {
        Task {
            await bootstrap()
            guard !disabled else { return }
            do {
                try await withMaintenanceSuspended {
                    let record = try await provider.add(url)
                    collections.append(NoteCollectionState(id: record.id, label: record.label))
                    operationError = nil
                    enqueue(record.id, due: .distantPast)
                }
            } catch {
                operationError = Self.message(error)
            }
        }
    }

    func remove(_ id: String) {
        Task {
            await bootstrap()
            guard !disabled else { return }
            do {
                try await withMaintenanceSuspended {
                    try await provider.remove(id)
                    collections.removeAll { $0.id == id }
                    pending.removeValue(forKey: id)
                    scans.remove(id)
                    failed.remove(id)
                    operationError = nil
                }
            } catch {
                operationError = Self.message(error)
            }
        }
    }

    func retry(_ id: String) {
        guard let collection = collections.first(where: { $0.id == id }),
              collection.status == .error || collection.status == .unavailable else {
            return
        }
        failed.remove(id)
        scans.insert(id)
        if pending[id] != nil {
            enqueue(id, due: .distantPast)
        }
        update(id) {
            $0.status = .pending
            $0.error = nil
            $0.progress = nil
        }
        pump()
    }

    func retrieve(query: [Float]) async throws -> [NotesHit] {
        var hits: [NotesHit] = []
        for collection in collections where collection.eligible {
            try Task.checkCancellation()
            do {
                hits += try await provider.search(collection.id, query: query)
            } catch NotesError.RebuildRequired {
                enqueue(collection.id, rebuild: true, due: .distantPast)
                update(collection.id) {
                    $0.status = .pending
                    $0.indexAvailable = false
                }
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                fail(collection.id, error)
            }
        }
        return hits
    }

    func verify(_ excerpts: [GroundedExcerpt]) async throws -> [GroundedExcerpt] {
        var result: [GroundedExcerpt] = []
        var accepted = 0
        for excerpt in excerpts {
            try Task.checkCancellation()
            guard case .localNote(let reference) = excerpt.source else {
                result.append(excerpt)
                continue
            }
            guard accepted < Int(limits.maxGroundingHits),
                  collections.contains(where: { $0.id == reference.collectionId }) else {
                continue
            }
            do {
                let reference = try await provider.verify(reference)
                result.append(GroundedExcerpt(
                    score: excerpt.score,
                    source: .localNote(reference: reference),
                    text: excerpt.text
                ))
                accepted += 1
            } catch is CancellationError {
                throw CancellationError()
            } catch NotesError.SourceChanged {
                enqueue(reference.collectionId, forced: [reference.documentId], due: .distantPast)
            } catch {
                fail(reference.collectionId, error)
            }
        }
        return result
    }

    func open(_ reference: NoteSourceReference) {
        Task {
            operationError = nil
            do {
                let url = try await withMaintenanceSuspended {
                    try await provider.preview(reference)
                }
                preview = NotesPreview(url: url)
                operationError = nil
            } catch NotesError.SourceChanged {
                operationError = "This note changed after the answer was generated."
            } catch NotesError.Unavailable {
                operationError = "This note is unavailable."
            } catch {
                operationError = "Could not open this note. Check folder access and try again."
            }
        }
    }

    private func enqueue(_ id: String, forced: Set<String> = [], rebuild: Bool = false, due: Date? = nil) {
        guard collections.contains(where: { $0.id == id }) else { return }
        let old = pending[id]
        generation &+= 1
        pending[id] = Pending(
            generation: generation,
            forced: (old?.forced ?? []).union(forced),
            rebuild: rebuild || old?.rebuild == true,
            due: due ?? old?.due ?? Date().addingTimeInterval(300)
        )
    }

    private func pump() {
        guard loaded, !disabled, foreground, scopes == 0, active == nil else { return }
        wake?.cancel()
        let embeddingReady = modelProvider.isEmbeddingModelReady()
        let request = embeddingReady ?
            pending.filter { !failed.contains($0.key) && $0.value.due <= Date() }.min { $0.key < $1.key } : nil
        guard let id = request?.key ?? scans.filter({ pending[$0] == nil && !failed.contains($0) }).min(),
              let collection = collections.first(where: { $0.id == id }) else {
            let due = embeddingReady ? pending.filter { !failed.contains($0.key) }.values.map(\.due).min() : nil
            let seconds = min(300, max(0.1, due?.timeIntervalSinceNow ?? 300))
            wake = Task { [weak self] in
                do {
                    try await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
                } catch {
                    return
                }
                guard let self else { return }
                self.scans.formUnion(self.collections.filter { $0.status != .unavailable &&
                    ($0.status != .error || $0.completedEmpty) }.map(\.id))
                self.pump()
            }
            return
        }
        scans.remove(id)
        let snapshot = request?.value
        let operation = NotesReadOperation()
        let cancellation = operation.cancellation
        let record = NotesCollectionInfo(id: collection.id, label: collection.label)
        let hadIndex = collection.indexAvailable
        let task = Task { @MainActor in
            defer {
                if active?.operation === operation {
                    active = nil
                }
                pump()
            }
            do {
                try cancellation.check()
                if let snapshot {
                    update(id) {
                        $0.status = hadIndex ? .updating : .indexing
                        $0.error = nil
                    }
                    let result = try await modelProvider.withEmbeddingContext(checkCancellation: cancellation.check) { context in
                        try await self.provider.index(
                            id,
                            context: context,
                            operation: operation,
                            options: NotesIndexOptions(
                                forcedDocumentIds: Array(snapshot.forced),
                                rebuild: snapshot.rebuild
                            )
                        ) { [weak self] progress in
                            Task { @MainActor in
                                guard let self, self.active?.operation === operation else { return }
                                self.update(id) {
                                    $0.progress = progress.percentage
                                    $0.documentCount = progress.indexedDocumentCount
                                }
                            }
                        }
                    }
                    let same = pending[id]?.generation == snapshot.generation
                    if same {
                        pending.removeValue(forKey: id)
                    }
                    if let changed = result.changedDocumentId {
                        enqueue(
                            id,
                            forced: Set(result.uncheckedDocumentIds).union([changed]),
                            due: same ? Date().addingTimeInterval(300) : nil
                        )
                    }
                    update(id) {
                        if pending[id] == nil {
                            $0 = Self.state(record, result.summary)
                        } else {
                            $0.status = .pending
                            $0.documentCount = result.summary.documentCount
                            $0.lastUpdatedAtMs = result.summary.lastUpdatedAtMs
                            $0.indexAvailable = hadIndex && result.summary.initialComplete && result.summary.documentCount > 0
                        }
                    }
                } else {
                    let result = try await provider.inspectFreshness(id, operation: operation)
                    update(id) {
                        $0 = Self.state(record, result.summary, progress: $0.progress)
                    }
                    if result.changed {
                        enqueue(id, forced: Set(result.forcedDocumentIds), due: result.summary.initialComplete ? nil : .distantPast)
                        update(id) {
                            $0.status = .pending
                            $0.error = nil
                        }
                    }
                }
            } catch NotesError.Cancelled {
                if snapshot == nil { scans.insert(id) }
                let saved = try? await provider.inspect(id)
                update(id) {
                    $0.status = .pending
                    if let saved { $0.lastUpdatedAtMs = saved.lastUpdatedAtMs }
                    $0.indexAvailable = hadIndex && saved?.initialComplete == true && (saved?.documentCount ?? 0) > 0
                }
            } catch NotesError.RebuildRequired {
                if snapshot == nil {
                    enqueue(id, rebuild: true, due: .distantPast)
                    update(id) {
                        $0.status = .pending
                        $0.indexAvailable = false
                    }
                } else {
                    fail(id, NotesError.RebuildRequired)
                }
            } catch {
                fail(id, error)
            }
        }
        active = Run(operation: operation, task: task)
    }

    private func fail(_ id: String, _ error: Error) {
        guard collections.contains(where: { $0.id == id }) else { return }
        failed.insert(id)
        scans.remove(id)
        update(id) {
            $0.status = (error as? NotesError) == .Unavailable ? .unavailable : .error
            $0.progress = nil
            $0.error = Self.message(error)
            $0.completedEmpty = false
        }
    }
    private static func message(_ error: Error) -> String {
        guard let error = error as? NotesError else { return "Could not update Your Notes. Please try again." }
        switch error {
        case .Unavailable: return "This Notes folder is unavailable. Check access and try again. If access has expired, remove the folder and add it again."
        case .SourceChanged: return "The folder changed. Retry to update Your Notes."
        case .SourceRead(let detail), .InvalidInput(let detail): return detail
        case .Storage: return "Could not save the Notes index. Check storage and try again."
        case .RebuildRequired: return "The Notes index needs rebuilding. Retry to update Your Notes."
        default: return "Could not update Your Notes. Please try again."
        }
    }

    private func update(_ id: String, _ body: (inout NoteCollectionState) -> Void) {
        guard let index = collections.firstIndex(where: { $0.id == id }) else { return }
        body(&collections[index])
    }
    private static func state(_ record: NotesCollectionInfo, _ summary: NotesSummary, progress: UInt8? = nil) -> NoteCollectionState {
        let ready = summary.initialComplete && summary.documentCount > 0
        let empty = summary.initialComplete && !ready
        return NoteCollectionState(id: record.id, label: record.label,
            status: ready ? .ready : summary.initialComplete ? .error : .pending,
            documentCount: summary.documentCount, lastUpdatedAtMs: summary.lastUpdatedAtMs, indexAvailable: ready,
            progress: summary.initialComplete ? nil : progress,
            error: empty ? emptyNotes : nil,
            completedEmpty: empty)
    }
}

struct NotesPreview: Identifiable {
    let id = UUID()
    let url: URL
}
